component/componentstatus/status.go

```diff
@@ -0,0 +1,128 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package componentstatus // import "go.opentelemetry.io/collector/component/componentstatus"
+
+import (
+	"time"
+)
+
+// Watcher is an extra interface for Extension hosted by the OpenTelemetry
+// Collector that is to be implemented by extensions interested in changes to component
+// status.
+type Watcher interface {
+	// ComponentStatusChanged notifies about a change in the source component status.
+	// Extensions that implement this interface must be ready that the ComponentStatusChanged
+	// may be called before, after or concurrently with calls to Component.Start() and Component.Shutdown().
+	// The function may be called concurrently with itself.
+	ComponentStatusChanged(source *InstanceID, event *Event)
+}
+
+type Status int32
+
+// Enumeration of possible component statuses
+const (
+	// StatusNone indicates absence of component status.
+	StatusNone Status = iota
+	// StatusStarting indicates the component is starting.
+	StatusStarting
+	// StatusOK indicates the component is running without issues.
+	StatusOK
+	// StatusRecoverableError indicates that the component has experienced a transient error and may recover.
+	StatusRecoverableError
+	// StatusPermanentError indicates that the component has detected a condition at runtime that will need human intervention to fix. The collector will continue to run in a degraded mode.
+	StatusPermanentError
+	// StatusFatalError indicates that the collector has experienced a fatal runtime error and will shut down.
+	StatusFatalError
+	// StatusStopping indicates that the component is in the process of shutting down.
+	StatusStopping
+	// StatusStopped indicates that the component has completed shutdown.
+	StatusStopped
+)
+
+// String returns a string representation of a Status
+func (s Status) String() string {
+	switch s {
+	case StatusStarting:
+		return "StatusStarting"
+	case StatusOK:
+		return "StatusOK"
+	case StatusRecoverableError:
+		return "StatusRecoverableError"
+	case StatusPermanentError:
+		return "StatusPermanentError"
+	case StatusFatalError:
+		return "StatusFatalError"
+	case StatusStopping:
+		return "StatusStopping"
+	case StatusStopped:
+		return "StatusStopped"
+	}
+	return "StatusNone"
+}
+
+// Event contains a status and timestamp, and can contain an error
+type Event struct {
+	status    Status
+	err       error
+	timestamp time.Time
+}
+
+// Status returns the Status (enum) associated with the Event
+func (ev *Event) Status() Status {
+	return ev.status
+}
+
+// Err returns the error associated with the Event.
+func (ev *Event) Err() error {
+	return ev.err
+}
+
+// Timestamp returns the timestamp associated with the Event
+func (ev *Event) Timestamp() time.Time {
```

Comment:
Timestamp seem to be use-less, because it is always "now", which means consumers (watchers) can also read if they need it.

Reply:
This is important for the healthcheck extension, which also needs to synthesize events. When synthesizing events the timestamp needs to be settable. We used to do this in core, but have removed it, at least temporarily. I was proposing we add an option to set the timestamp in a followup: https://github.com/open-telemetry/opentelemetry-collector/pull/10730#discussion_r1696051237.

Reply:
What about, if this is needed for that extension, we can have `StatusWithTime struct {Status, Time}` in the healthcheck and not here. No point in having it here.

Reply:
Also, until we discuss the next PR, I would rather remove this for the moment.

Reply:
I prefer to keep this PR a 1-to-1 move of stuff related to component status reporting into `componentstatus`.  In my opinion we can have API discussions about `componentstatus` in future issues/PRs. I can update the godocs to state that everything in `componentstatus` is experimental and currently exempt from our deprecation process if you'd like.

Reply:
> I prefer to keep this PR a 1-to-1 move of stuff related to component status reporting into componentstatus. In my opinion we can have API discussions about componentstatus in future issues/PRs. I can update the godocs to state that everything in componentstatus is experimental and currently exempt from our deprecation process if you'd like.

Then update the title to "move code" instead of adding a new module which to me means we can do a better job.

Reply:
I think a timestamp will be useful for other status watchers and worth considering for the generated status event. Otherwise, if there are multiple watchers that care about the timestamp, each one will have to wrap the event in a custom struct and generate timestamps independently.

Reply:
> I think a timestamp will be useful for other status watchers and worth considering for the generated status event. Otherwise, if there are multiple watchers that care about the timestamp, each one will have to wrap the event in a custom struct and generate timestamps independently.

When in doubt leave it out. I am not convinced yet, and until we have 2-3 use-cases we should not add it, since adding later is backwards compatible, but currently we pay the cost of reading the time as well as the maintenance cost.

Reply:
Followup issue created: https://github.com/open-telemetry/opentelemetry-collector/issues/10763

Reply:
I will still suggest the timestamp has value. I wrote the following on the issue:

> I believe that we should keep the timestamp as the observed time is not the same time as the observation. Since status watchers are dispatched in sequence, any status watcher that does a non-trivial amount of work, or something that could block (like writing to an event to a channel) could delay the observed time for subsequent watchers. This would be especially problematic if the delay was variable between invocations. Currently the timestamp is used by the health check extension for time calculations.